### ISSUE TITLE
Bug with buttons inside headings with accessible names

### ISSUE THREAD

**Issue by scottaohara** [nvaccess/nvda#13296]
Ran into this issue when encountering a button inside of a `div role=heading` with an `aria-label` specified on the heading.... which was unnecessary... but technically it is valid per the ARIA specification.  In this instance, the button was not consistently discoverable when navigating the page with scan mode, and activating the button was not possible unless I switched to application mode.

Narrator and JAWS did not experience this issue when testing.

All tests were performed with latest Edge, Chrome and Firefox.

### Steps to reproduce:
Go to the following test page: https://codepen.io/scottohara/full/jOaqMKq
Attempt to activate the "test case" button while NVDA is in scan mode.

### Actual behavior:
Cannot activate the 'test case' button in scan mode.  There are additional issues with attempting to find the button when navigating with scan mode or using the `b` key to navigate by buttons.

### Expected behavior:
Should be able to activate and find the button without issue. 

#### NVDA version: 
2021.3 and 2021.3.1

#### Windows version:
10

#### Does the issue still occur after restarting your computer?
yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
noted above

#### If NVDA add-ons are disabled, is your problem still occurring?
n/a

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
n/a

**Comment by feerrenrut** [comment]
I haven't investigated this yet, but a note that may be relevant:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/heading_role#accessibility_concerns
>Warning: Using aria-label or aria-labelledby will hide the content of your heading from assistive technologies, reading the label instead of the heading.

**Comment by scottaohara** [comment]
Yeah. Here be a reason why that warning was written then I guess. 

Please note, I’m not saying I agree with this pattern. I never much understood or agreed with naming a heading in this way, rather than by its contents. But this is conforming to what the aria spec says is allowed. For better or worse.

**Comment by Adriani90** [comment]
Is there any update on this one?
cc: @jcsteh this also impacts Firefox.
Tested with NVDA 2025.3.1.

**Comment by jcsteh** [comment]
I don't really know what NVDA is expected to do in this case. Authors complained for a long time that aria-label was ignored on things, so NVDA was changed to override the content for headings, buttons, etc. But a case like this explicitly requires that the aria-label does not override the content. So what's the expected behaviour here? Include both the aria-label and the content, potentially causing weird duplication? This is exactly the sort of nonsense I was concerned about when this change was made years ago, as I outlined in [this blog post](https://www.jantrid.net/2015/12/woe-aria-surprisingly-but-ridiculously.html).

**Comment by Adriani90** [comment]
Thanks Jamie for your valuable info.
Is it possible to allow the interaction in browse mode though? The risk is that the users hear the aria-label instead of the content which might be the wrong label of a button, but this could be mittigated if NVDA optionally announces something like "has text" when landing on the button.

It seems NVDA can already distinguish whether a button has only content or both aria-label and content.
So in this case, wehn using e.g. arrow keys in browse mode, NVDA would announce
"heading level 2 button has text not pressed, foo"

where "has text" means aria label has overwritten the content.
A key stroke could be implemented which allows users to hear the true content, or at least people get aware of it and switch to focus mode to hear the true content on demand.
This could also work for any other element where NVDA ignores content in favour of aria-label.

**Comment by jcsteh** [comment]
1. There are many, many controls that override the content with aria-label. You'd be hearing "has text" all over the place.
2. There could be a lot more than just a single button inside a heading. There could be an entire section for that matter, including multiple interactive controls. I don't see a way to make that sensibly navigable or report it sensibly.

What does JAWS do (specifically) in this situation? I'm fairly sure any solution here is going to cause its own set of problems that make other use cases worse, but it can't hurt to have info on how other tools handle this.

**Comment by Adriani90** [comment]
Testing with Jaws 2025, Jaws ignores aria-label and keeps the true content in browse mode, and it is possible to interact with the test case button.

**Comment by scottaohara** [comment]
FWIW, this sort if issue is exactly why I’ve been proposing in aria that headings shouldn’t allow name from author, among some other roles that seem to be footguns like this.

**Comment by jcsteh** [comment]
That makes me wonder whether JAWS just ignores aria-label on headings altogether:
`data:text/html,before<h1 aria-label="label">content</h1>after`
Does JAWS report "content" or "label" for the heading?

**Comment by Adriani90** [comment]
Jaws reports label, same as NVDA.

**Comment by jcsteh** [comment]
Okay. What about these?

1. `data:text/html,a <h1 aria-label="b"><button aria-label="c">d</button> e</h1>`
2. `data:text/html,a <h1 aria-label="b">c<button aria-label="d">e</button>f</h1> g`
3. `data:text/html,a <h1 aria-label="b">c <button aria-label="d">e</button> f <button aria-label="g">h</button> i</h1> j`

**Comment by jcsteh** [comment]
Oo, and this:

4. `data:text/html,a <h1 aria-label="b"><div>c</div><div>d <button aria-label="e">f</button></div></h1>g`

**Comment by jcsteh** [comment]
This is only the wildness I can come up with off the top of my head without trying particularly hard. There are likely a bunch of even more problematic cases.

**Comment by Adriani90** [comment]
Hmm, in all these cases Jaws ignores the aria-label of the headings, but not the aria-label of the buttons. It ignores however the content of the buttons if there is an aria-label for the button itself.

1. Jaws reports:
a
c Heading level 1 button
heading level 1 e

2. Jaws reports:
a
Heading level 1 c
d Heading level 1 button
Heading level 1 f
g

3. Jaws reports:
a
Heading level 1 c
d heading level 1 button
heading level 1 f
g heading level 1 button
heading level 1 i
j

4. Jaws reports:
a
heading level 1 c
Heading level 1 d
e heading level 1 button
g

**Comment by Adriani90** [comment]
Except for this, why does NVDA ignore the button completely when its id refers to the aria-label of a heading? Actually this should impact only the label as such, but not the ```<button>``` element as such.

**Comment by jcsteh** [comment]
Thanks. Okay, two more:

5. `data:text/html,a <h1 aria-label="b">c<span>d</span></h1> e`
6. `data:text/html,a <h1 aria-label="b">c <strong>d</strong></h1> e`

**Comment by Adriani90** [comment]
Confusing or not, Jaws and NVDA behave the same in test cases 5 and 6. They report:
a
Heading level 1 b
e

It seems Jaws ignores the aria-label on headings only when an object with interaction is inside the ```<h>``` tags.

See following example, which makes it even more significant:

```
data:text/html,a <h1 aria-label="b"><a aria-label="c" title="d" href="">e</a> f</h1>
```

NVDA ignores the link completely in browse mode and reports:
a
heading level 1 b

Jaws on the other hand ignores aria-label on the heading and also content of ```<a>``` and reports:
a
heading level 1 link c d
heading level 1 f

In my view it is more desirable what jaws does in this case, but it is ofcourse probably debatable as well.

**Comment by jcsteh** [comment]
Fair enough. So the rule is: if the heading contains interactive content, ignore the aria-label and render the content. That makes sense, putting aside the fact that it's a horrible, inconsistent hack.

Unfortunately, this is not going to be easy to implement as things currently stand. Filling vbuf backends (fillVBuf) is currently recursive: they take a node, then they render all its children, etc. So, to do this, we'd have to:

1. Render the heading itself.
2. Render all its descendants, which could be thousands of nodes for all we know.
3. When we return to the heading, somehow figure out whether there were interactive descendants.
    - We don't want to walk the entire subtree again to figure this out. That's too expensive and redundant, especially if there are thousands of nodes.
    - There isn't currently any way to communicate back up the stack whether there were any interactive descendants. Maybe we need some sort of state stack that we pass around.
    - We can't just set a flag on the heading while rendering the descendants, since the heading might be a parent, a grandparent, a great grandparent, etc. We absolutely don't want to do an ancestor walk for every node we render.
    - Maybe we could set a flag on any parent that contains interactive descendants, regardless of its ancestry, but that feels like a horrible waste of memory, since we only need this for headings.
4. If there were interactive descendants, we've already rendered them, so we're done; we can all go home.
5. If not, we have to remove all the content that was just rendered and render the name (aria-label) instead.
    - No vbuf backend currently removes nodes while rendering. We have VBufStorage_buffer_t::removeFieldNode, but that's only used by the base code when handling actual removals. I have no idea whether there are issues with calling this from fillVBuf. If there are, those will need to be addressed.

**Comment by Adriani90** [comment]
> Fair enough. So the rule is: if the heading contains interactive content, ignore the aria-label and render the content.

I would formulate the rule rather like:
if the heading contains interactive content, ignore the aria-label of the heading and render the content, or aria-label and title (if any) of the interactive element instead.

Yeahm the implementation seems challenging though, thanks alot for this valuable explanation.

**Comment by JAWS-test** [comment]
Following @scottaohara attempt in 2022, I reminded the ARIA Working Group once again that the issue needs to be resolved in the specification: https://github.com/w3c/aria/issues/2797
